### ISSUE TITLE
Make it possible to set destination filename

### ISSUE THREAD

**Issue by probonopd** [probonopd/linuxdeployqt#146]
From https://github.com/LMMS/lmms/pull/3688#issuecomment-313926775

> Can you please add an equivalent of the appimagetool DESTINATION parameter to the -appimage flag of linuxdeployqt or simply add a flag to stop changing the name? If not, we'll be forced to either leave a lingering shellcheck FIXME or we can stop using linuxdeployqt for creating the image.

**Comment by probonopd** [comment]
Let's discuss about this rationale:

> A project should be able to explicitly control the file name without trying to guess. 

`linuxdeployqt` is supposed to produce reproducible file names using this scheme:

APPNAME-ARCH.AppImage

* APPNAME comes from the `Name=` entry of the desktop file
* ARCH is determined by inspecting the contents of the AppDir

**Comment by TheAssassin** [comment]
Well, the pragmatic solution is to use appimagetool in a second call, because you can pass the AppImage name as a parameter.

But I like the idea of passing a name to linuxdeployqt as well. How about allowing `linuxdeployqt -appimage=myapp.AppImage`, and keep falling back to the automagic solution that is currently implemented if there's no value passed for `-appimage` as it is currently implemented?

**Comment by probonopd** [comment]
I'd welcome a PR if you think it's useful.

**Comment by TheAssassin** [comment]
It is useful, but has a lesser priority for me. I'd encourage anyone who needs this feature to send a PR.

**Comment by dalboris** [comment]
Just commenting to say that I am interested in this feature too. I may try to implement it and send a PR if/when time allows :-)

The current naming scheme may indeed be reproducible, but reproducing this naming scheme on client side (or using a glob-pattern to locate the output) feels a bit too hacky and unreliable for comfort. Most commands generating output have a way to specify the name and location of the output.

And in fact, I think the current default name and location isn't the best it could be (although it's probably not a good idea to change it now, since some scripts may rely on it). For example, if we do:

```
linuxdeployqt ./subdir/foo-2.0-debug-64bit.appdir/usr/share/applications/foo.desktop -appimage
```

Then currently, assuming `Name=Foo` in the desktop entry, the appimage is created at:

```
./Foo-x86_64.AppImage
```

Note how the current working directory affects where the appimage is created, thus the AppImage isn't generated alongside its corresponding appdir which is under `./subdir`. Also, the "2.0" version info is lost, together with the stylistic choice to use "64bit" instead of "x86_64", and other variant info such as "debug". I would say that a better default name and location would in this case be:

```
./subdir/foo-2.0-debug-64bit.AppImage`
```

(but again, I wouldn't advocate on making this change, for backward compatibility).

Anyway, thanks for making this great tool, it actually works better than macdeployqt, for which I've ran into a few nasty bugs.

**Comment by probonopd** [comment]
Please don't use "64bit". It is not clear what architecture "64bit" means.
The architecture is called "x86_64" for Intel, and "aarch64" for ARM.

By diverging from the standard file naming scheme, things generally become more troublesome and more complicated.

**Comment by dalboris** [comment]
Fair enough. This was in fact just an example, but it raises an interesting point: my planned naming scheme was actually to skip the architecture altogether:

```
vgcillustration-alpha-2019-10-11.1.AppImage
```

since I am not planning to support 32bit processors anyway (i.e., no need to differentiate between x86 and x86_64). But I suppose I may want to support the use case of using my app on Linux on a tablet/mobile in the future, in which case I would indeed have to differentiate between x86_64 and aarch64.

I don't have any experience building for ARM, do you know of good examples of projects deploying AppImages for ARM? Is it even possible to target ARM using TravisCI, for example? Well, actually I just looked at the doc, and there is just some experimental support since less than a month:

https://blog.travis-ci.com/2019-10-07-multi-cpu-architecture-support
https://docs.travis-ci.com/user/multi-cpu-architectures/

So I guess there aren't yet a lot of project using TravisCI to deploy ARM AppImages.

**Comment by TheAssassin** [comment]
We're cross-building the AppImageKit binaries (mainly appimagetool) manually using qemu emulation in a CentOS 6 container. This is of course a lot easier to do.

The name of your AppImage should still show what architecture it aims for. You probably underestimate the amount of 32-bit users (or people on RasPis seeking to run apps as AppImages).

**Comment by dalboris** [comment]
About not supporting 32bit users: my app is quite GPU-intensive, similar to, say, Blender. I don't see any use case where someone would have the required CPU/GPU power without a >=64bit processor ;-)

**Comment by TheAssassin** [comment]
I'm not saying you should support 32-bit users. I'm saying you should *signalize* those users that the AppImage won't run there. Otherwise, the AppImage will just crash silently and they're left scratching their head.

In some future release, [AppImageLauncher](https://github.com/TheAssassin/AppImageLauncher) is going to detect these compat issues and will warn about those problems. But we're not there yet.

**Comment by dalboris** [comment]
Oh, I see, I didn't get that. Yes, that makes sense.

On the other hand, adding the architecture as per AppImage convention does create inconsistencies between the names of the AppImages vs. the names of the macOS / Windows bundles and installers,  where I don't specify the architecture because they either don't support 32bit since forever (macOS), or have built-in support to detect compat issues (Windows MSI). Such inconsistencies between naming schemes make things a little more complicated on my side (server scripts), and perhaps for a few rare advanced users who'd like to automate installation on multiple platforms.

I think I'll follow the AppImage convention anyway to try to be as idiomatic as possible for each platform. I'm just raising a few points for completeness sake: you can't have per-app name consistency across all platforms, and per-platform name consistency across all apps.

**Comment by TheAssassin** [comment]
A cynical person might comment that if your naming convention doesn't provide that information it should be extended on all platforms to be consistent again...

**Comment by dalboris** [comment]
Indeed, but a marketing person might argue that Windows users really like "VGC Illustration 2020 Installer.exe" much better than "vgcillustration-2020-x86_64-installer.exe", and that they represent >80% of our user base ;-) Anyway, we digress, although it was a useful discussion to raise my awareness on some important points.

I suppose we do agree that it would still be useful to explicitly specify the output name, even if we follow the current default naming convention. It allows to specify the parent directory where the AppImage is to be created (if different from the current working directory), and it makes client scripts more robust to potential changes in linuxdeployqt default naming scheme. Currently, when a script needs to do some further actions on the generated AppImage (typically, uploading it somewhere), then it does have to "guess" the output name chosen by linuxdeployqt/appimagetool (where "guess" means "recompute the name using the exact same non-officially documented naming scheme"), and hope to get it right. This is neither trivial nor good practice. Regex or glob patterns can't work on all use cases either, since we may want to generate several variants of AppImages but only post-process or upload some of them.

**Comment by probonopd** [comment]
ARM 32-bit (armv7) and ARM 64-are becoming very important with Chromebooks and boards and notebooks like Raspberry Pi and Pinebook becoming more and more common.

**Comment by tresf** [comment]
> A cynical person might comment that if your naming convention doesn't provide that information it should be extended on all platforms to be consistent again...

As OP of this issue, I feel the need to add my own experiences... 

I help maintain two packages... 

1. Java - We don't specify architecture in the file name (to be fair, we don't use `.AppImage` at all, but instead use `makeself` with some platform-agnostic shell commands) 
2. C++ - We do specify architecture in the file name.

Regarding Mac, it's [widely rumored](https://www.tomshardware.com/news/apple-mac-arm-cpus-2020-intel,38668.html) that they'll be offering an ARM64, bringing back the "universal package" problem they had in the PPC, 32-bit Intel, 64-bit Intel transition days.

Regarding Linux, especially with the rise of ChromeBooks supporting the `Linux (Beta)` feature, ARM is slowly becoming the most requested platform to support from our C++ app.  Note, there are serious limitations to running these (network, USB, etc) but the adoption rate is something worth considering. 

Regarding Windows, [it too supports ARM](https://docs.microsoft.com/en-us/windows/arm/), so this distinction may be worthwhile.

In regards to how you present this to the user, I can reassure you that as long as you present the correct download button to your users, they won't mind a few extra numbers and letters in your binary name. :D
